File: glenn/121/ruby/max_profit.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  max_profit = 0

  begin_val = prices[0]
  end_val = prices[0]

  for idx in 1..prices.count - 1
    val = prices[idx]
    if val < begin_val
      # We've encountered a new possible profit span
      profit = end_val - begin_val
      max_profit = profit if profit > max_profit
      begin_val = val
      end_val = val
    elsif val > end_val
      end_val = val
    end
  end
  profit = end_val - begin_val
  max_profit = profit if profit > max_profit

  max_profit
end

# Runtime: 136 ms, faster than 92.66% of Ruby online submissions for Best Time to Buy and Sell Stock.
# Memory Usage: 219.5 MB, less than 62.08% of Ruby online submissions for Best Time to Buy and Sell Stock.

File: glenn/121/go/max_profit.go
package glenn

func maxProfit(prices []int) int {
	mxProfit := 0
	beginVal := prices[0]

	for i := 1; i < len(prices); i++ {
		if prices[i] < beginVal {
			beginVal = prices[i]
		} else if prices[i] - beginVal > mxProfit {
			mxProfit = prices[i] - beginVal
		}
	}

	return mxProfit;
}

// Runtime: 120 ms, faster than 96.40% of Go online submissions for Best Time to Buy and Sell Stock.
// Memory Usage: 8.3 MB, less than 97.00% of Go online submissions for Best Time to Buy and Sell Stock.
